include/deal.II/lac/petsc_snes.templates.h

```diff
@@ -69,12 +69,16 @@ namespace PETScWrappers
           eptr = nullptr;
           return 0;
         }
-      // In case of an exception, capture the exception and
-      // return -1:
+      catch (const RecoverableUserCallbackError &)
+        {
+          // ok, ignore, but reset the pointer
+          eptr = nullptr;
+          return -1;
+        }
       catch (...)
         {
           eptr = std::current_exception();
-          return -1;
+          return -2;
         }
     }
```

Comment:
Out of curiosity: Does PETSc interpret return values of -1 and -2 differently?

Reply:
no, it is me in the calling site interpreting it differently. With this patch, I always return `PETSC_ERR_LIB` (error in external library) to PETSc